lib/diskoverflow/map.go

```diff
@@ -0,0 +1,315 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/syndtr/goleveldb/leveldb"
+	"github.com/syndtr/goleveldb/leveldb/iterator"
+	"github.com/syndtr/goleveldb/leveldb/opt"
+)
+
+type Map interface {
+	Set(k string, v Value)
+	Get(k string, v Value) bool
+	Pop(k string, v Value) bool
+	Delete(k string)
+	NewIterator() MapIterator
+	Bytes() int
+	Items() int
+	SetOverflowBytes(bytes int)
+	Close()
+}
+
+type omap struct {
+	commonMap
+	base
+}
+
+type commonMap interface {
+	common
+	set(k string, v Value)
+	Get(k string, v Value) bool
+	Pop(k string, v Value) bool
+	Delete(k string)
+	newIterator(p iteratorParent) MapIterator
+}
+
+// NewMap returns an implementaiton of Map, spilling to disk at location.
+func NewMap(location string) Map {
+	o := &omap{base: newBase(location)}
+	o.commonMap = &memoryMap{
+		values: make(map[string]Value),
+	}
+	return o
+}
+
+func (o *omap) Set(k string, v Value) {
+	if o.iterating {
+		panic(concurrencyMsg)
+	}
+	if o.startSpilling(o.Bytes() + v.ProtoSize()) {
+		d, err := v.Marshal()
+		errPanic(err)
+		newMap := newDiskMap(o.location)
+		it := o.newIterator(o)
+		for it.Next() {
+			v.Reset()
+			it.Value(v)
```

Comment:
Value perhaps should do a reset? I guess this is internal, hence it does not matter, but users are bound to forget that (unless the reflect magic means it's not necessery)